File: csharp/game-of-life/GameOfLife.cs
public static class GameOfLife
{
    public static int[,] Tick(int[,] matrix)
    {
        int rows = matrix.GetLength(0);
        int cols = matrix.GetLength(1);

        var nextGen = new int[rows, cols];
        for (int row = 0; row < rows; row++)
        {
            for (int col = 0; col < cols; col++)
            {
                int liveNeighbors = CountLiveNeighbor(matrix, row, col);
                bool isAlive = matrix[row, col] == 1;

                if (isAlive && (liveNeighbors == 2 || liveNeighbors == 3))
                {
                    nextGen[row, col] = 1;
                }
                else if (!isAlive && liveNeighbors == 3)
                {
                    nextGen[row, col] = 1;
                }
            }
        }
        return nextGen;
    }
    private static int CountLiveNeighbor(int[,] matrix, int row, int col)
    {
        int rows = matrix.GetLength(0);
        int cols = matrix.GetLength(1);
        int liveCount = 0;
        for (int i = -1; i <= 1; i++)
        {
            for (int j = -1; j <= 1; j++)
            {
                if (i == 0 && j == 0) continue;
                int neighborRow = row + i;
                int neighborCol = col + j;
                if (neighborRow >= 0 && neighborRow < rows &&
                    neighborCol >= 0 && neighborCol < cols)
                {
                    if (matrix[neighborRow, neighborCol] == 1)
                    {
                        liveCount++;
                    }
                }
            }
        }
        return liveCount;
    }
}

File: javascript/game-of-life/game-of-life.js
//
// This is only a SKELETON file for the 'Conway's Game of Life' exercise. It's been provided
// as a convenience to get you started writing code faster.
//

export class GameOfLife {
  constructor(matrix = []) {
    this.matrix = matrix;
  }

  tick() {
    if (this.matrix.length === 0) return;

    const rows = this.matrix.length;
    const cols = this.matrix[0].length;
    const nextGeneration = Array.from({ length: rows }, () => Array(cols).fill(0));

    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        const liveNeighbors = this.countLiveNeighbors(r, c, rows, cols);
        const cell = this.matrix[r][c];

        if (cell === 1 && (liveNeighbors === 2 || liveNeighbors === 3)) {
          nextGeneration[r][c] = 1;
        } else if (cell === 0 && liveNeighbors === 3) {
          nextGeneration[r][c] = 1;
        }
      }
    }
    this.matrix = nextGeneration;
  }
  
  countLiveNeighbors(row, col, rows, cols) {
    const directions = [
      [-1, -1], [-1, 0], [-1, 1],
      [0, -1],           [0, 1],
      [1, -1],  [1, 0],  [1, 1]
    ];

    let count = 0;
    for (const [dr, dc] of directions) {
      const nr = row + dr;
      const nc = col + dc;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
        count += this.matrix[nr][nc];
      }
    }
    return count;
  }

  state() {
    return this.matrix;
  }
}
